zebra-chain/src/value_balance.rs

```diff
@@ -0,0 +1,248 @@
+    AmountError(#[from] crate::amount::Error),
+}
+
+impl<C> std::ops::Add<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn add(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent + rhs.transparent)?,
+            sprout: (value_balance.sprout + rhs.sprout)?,
+            sapling: (value_balance.sapling + rhs.sapling)?,
+            orchard: (value_balance.orchard + rhs.orchard)?,
+        })
+    }
+}
+
+impl<C> std::ops::Sub<ValueBalance<C>> for Result<ValueBalance<C>, ValueBalanceError>
+where
+    C: Constraint,
+{
+    type Output = Result<ValueBalance<C>, ValueBalanceError>;
+
+    fn sub(self, rhs: ValueBalance<C>) -> Self::Output {
+        let value_balance = self?;
+
+        Ok(ValueBalance::<C> {
+            transparent: (value_balance.transparent - rhs.transparent)?,
+            sprout: (value_balance.sprout - rhs.sprout)?,
+            sapling: (value_balance.sapling - rhs.sapling)?,
+            orchard: (value_balance.orchard - rhs.orchard)?,
+        })
+    }
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+use proptest::prelude::*;
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NegativeAllowed> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+            any::<Amount>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(any(test, feature = "proptest-impl"))]
+impl Arbitrary for ValueBalance<NonNegative> {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+            any::<Amount<NonNegative>>(),
+        )
+            .prop_map(|(transparent, sprout, sapling, orchard)| Self {
+                transparent,
+                sprout,
+                sapling,
+                orchard,
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    proptest! {
+        #[test]
+        fn test_operations(
+            value_balance1 in any::<ValueBalance>(),
+            value_balance2 in any::<ValueBalance>())
+        {
+            zebra_test::init();
+
+            // test the add operator
+            let add = Ok(value_balance1) + value_balance2;
+
+            if !add.is_err() {
```

Comment:
Idea: for the check, not sure if it's too complicated:

```
let transparent = value_balance1.transparent + value_balance2.transparent;
let sprout = value_balance1.sprout + value_balance2.sprout;
let sapling = value_balance1.sapling + value_balance2.sapling;
let orchard = value_balance1.orchard + value_balance2.orchard;

match (transparent, sprout, sapling, orchard) {
    (Ok(transparent), Ok(sprout), Ok(sapling), Ok(orchard)) => prop_assert_eq!(
        value_balance1 + value_balance2,
        Ok(ValueBalance {
            transparent,
            sprout,
            sapling,
            orchard,
        })
    ),
    _ => prop_assert_eq!(
        value_balance1 + value_balance2,
        Err(ValueBalanceError::Amount(amount::Error::??))
    ),
}
```

Reply:
I think this a bit better than what we had so implemented it https://github.com/ZcashFoundation/zebra/pull/2505/commits/4b5f1e64489ecc90d81c64411f8ffa0e6880c97f

What i was not able to do is to get rid of the string error matching.

Reply:
I'll see if I can get the error matching working, it's important to get good test failure output.

It could be tricky, because the proptest macros disable some LSP features and compiler hints.

Reply:
I made a suggestion in the code to use `matches!`